pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,192 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build kubeapiserver
+
+package leaderelection
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+	log "github.com/cihub/seelog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/leaderelection"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+const (
+	defaultLeaderLeaseDuration = 60 * time.Second
+	defaultLeaseName           = "datadog-leader-election"
+	clientTimeout              = 2 * time.Second
+)
+
+var (
+	globalHolderIdentity      string
+	globalLeaderEngine        *LeaderEngine
+	globalLeaderLeaseDuration = 0 * time.Second
+)
+
+// LeaderEngine is a structure for the LeaderEngine client to run leader election
+// on Kubernetes clusters
+type LeaderEngine struct {
+	initRetry retry.Retrier
+
+	once sync.Once
+
+	HolderIdentity string
+	LeaseDuration  time.Duration
+	LeaseName      string
+	coreClient     *corev1.CoreV1Client
+	leaderElector  *leaderelection.LeaderElector
+
+	currentHolderIdentity string
+	currentHolderMutex    sync.RWMutex
+}
+
+func newLeaderEngine() *LeaderEngine {
+	return &LeaderEngine{
+		LeaseName: defaultLeaseName,
+	}
+}
+
+// ResetGlobalLeaderEngine is a helper to remove the current LeaderEngine global
+// It is ONLY to be used for tests
+func ResetGlobalLeaderEngine() {
+	globalLeaderEngine = nil
+}
+
+// SetLeaderLeaseDuration is a helper to set the current LeaderLeaseDuration global
+// It is ONLY to be used for tests
+func SetLeaderLeaseDuration(ttl time.Duration) {
```

Comment:
**nit:** Instead of having `SetLeaderLeaseDuration, SetHolderIdentify, globalLeaderLeaseDuration, globalLeaderLeaseDuration` , could identity and leaseDuration be passed as parameters of a `GetCustomLeaderEngine` that would keep these as class members?

`GetLeaderEngine` could then just call `GetCustomLeaderEngine` with `""` and `defaultLeaderLeaseDuration`